README.md
# shale
An elegant and extensible reverse-polish notation programming language.

Elegant because of its power and simplicity.

Extensible through libraries. Currently these include but are not limited to

* a thread library to create threads, semaphores and mutexes
* a maths library
* a library dedicated to prime numbers.

Compiles on Linux and Mac OS X. To compile, do

* ./configure; make install

See shale -h and shale -s for details, and the examples/ directory.


ChangeLog
shale:

  1.3.16 - 20 Nov 2022
    - add += -= *- and /= operators

  1.3.15 - 28 Oct 2021
    - include support for entering binary numbers with a 0b prefix.

  1.3.14 - 03 Jul 2021
    - support for the namespace library

  1.3.13 - 29 Jun 2021
    - improved thread performance by using a mutex cache

  1.3.12 - 29 Jun 2021
    - fix the incorrect release of objects back into the cache when using threads

  1.3.11 - 28 Jun 2021
    - improved thread performance by using one object cache per thread,
      eliminating the single cache and mutex, along with its inherent mutex
      contention, across all threads

  1.3.10 - 27 Jun 20221
    - improved speed and lower memory use

  1.3.9 - 25 Jun 2021
    - improved thread performance
    - improved error message for unknown characters

  1.3.8 - 31 May 2021
    - removed several compiler warnings
    - improved array and thread library error message

  1.3.7 - 28 May 2021
    - improved thread performance

  1.3.6 - 17 Apr 2021
    - stack operator now displays pointer values rather than bailing

  1.3.5 - 16 Apr 2010
    - the value operator now handles pointers

  1.3.4 - 11 Apr 2021
    - numbers now use lass memory

  1.3.3 - 9 Apr 2021
    - fix zero pointer dereference in the ++ and -- operators when the variable
      is defined but not initialised

  1.3.2 - 5 Mar 2021
    - add support for the string library

  1.3.1 - 23 Feb 2021
    - improved performance of threaded scripts

  1.3.0 - 22 Feb 2021
    - shalelib is now thread-safe
    - use "Error" rather than "Internal error" for exceptions raised
      where no lexical information is available

  1.2.6 - 21 Jan 2021
    - use the cache to allocate some static numbers (version nubmers)
      and strings at startup

  1.2.5 - 17 Jan 2021
    - addition of the ++ and -- operators

  1.2.4 - 11 Jan 2021
    - addition of language option:: shale::

  1.2.3 - 10 Jan 2021
    - addition of left shift (<<) and right shift (>>) operators



namespace library:

  1.0.0 - 03 Jul 2021
    - initial release
    - shale version 1.3.14



string library:

  1.0.2 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  1.0.1 - 25 Mar 2021
    - initial release of atoi and atof functions
    - shale version 1.3.2

  1.0.0 - 05 Mar 2021
    - initial release
    - shale version 1.3.2



thread library:

  1.0.7 - 02 Jul 2021
    - use a variable's name rather than its value when dealing with mutexes and semaphores
    - shale version 1.3.13

  1.0.6 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  1.0.5 - 27 Jun 2021
    - improved thread safety
    - shale version 1.3.9

  1.0.4 - 31 May 2021
    - removed some compiler warnings
    - improves error messages
    - shale version 1.3.8

  1.0.3 - 28 May 2021
    - changed support for the BTree and Cache classes

  1.0.2 - 23 Feb 2021
    - support improved performance of threaded scripts
    - shale version 1.3.1

  1.0.1 - 22 Feb 2021
    - add argument to create thread::()
    - shalelib is now thread-safe
    - shale version 1.3.0
     
  1.0.0 - 21 Feb 2021
    - initial release
    - shale version 1.2.6



primes library:

  1.2.2 - 28 Jun 2021
    - use new cache model
    - hale version 1.3.11

  1.2.1 - 27 Jun 2021
    - use of the lower memory footprint true and false values
    - shale version 1.3.10

  1.2.0 - 18 Feb 2021
    - add Euler's totient functon: phi primes::()
    - shale version 1.2.6

  1.1.1 - 17 Feb 2021
    - fix loop termination condition on sieve type generate
    - shale version 1.2.6

  1.1.0 - 31 Jan 2021
    - Initial release of a sieve layout type
    - shale version 1.2.6

  1.0.0 - 10 Jan 2021
    - initial release
    - shale version 1.2.2

 

time library:

  1.1.3 - 18 Jul 2021
    - fix to the localtime time::() function
    - shale version 1.3.14

  1.1.2 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  1.1.1 - 22 Feb 2021
    - add sleep time::()
    - shale version 1.3.0

  1.1.0 - 17 Jan 2021
    - add dateformat time:: variable to control output format
      of date time::()
    - shale version 1.2.5

  1.0.3 - 11 Jan 2021
    - change localtime time::() to store entries in tm:: time::,
      for example sec tm:: time::
    - shale version 1.2.4

  1.0.2 - 11 Jan 2021
    - added timems time::() to include milliseconds
    - added localtime time::() to return contents of struct tm
    - shale version 1.2.4

  1.0.1 - 11 Jan 2021
    - language support (en, de, fr) in date time::()
    - shale version 1.2.4

  1.0.0 - 10 Jan 2021
    - initial release
    - shale version 1.2.3



maths library:

  - 1.0.4 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  - 1.0.3 - 31 Mar 2021
    - initial release of random and srandom
    - shale release 1.3.2

  - 1.0.2 - 1 Feb 2021
    - initial release of a GCD (HCF) function
    - shale version 1.2.6

  - 1.0.1 - 21 Jan 2021
    - use the cache to allocate some static numbers (version nubmers)
      and strings at startup
    - shale version 1.2.6



file library:

  1.0.3 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  1.0.2 - 21 Jan 2021
    - use the cache to allocate some static numbers (version nubmers)
      and strings at startup
    - shale version 1.2.6



array library:

  1.0.3 - 28 Jun 2021
    - use new cache model
    - shale version 1.3.11

  1.0.2 - 31 May 2021
    - removed some compiler warnings
    - improved error messages

  1.0.1 - 21 Jan 2021
    - use the cache to allocate some static numbers (version nubmers)
      and strings at startup
    - shale version 1.2.6


src/shale.h
/*

MIT License

Copyright (c) 2020-2021 Graeme Elsworthy <github@sharkshead.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include "shalelib.h"

#define MAJOR ((INT)  1)
#define MINOR ((INT)  3)
#define MICRO ((INT) 16)

// Lexical analyser stuff.

// The tokens returned.
#define LEX_TOKEN_EOF                       0
#define LEX_TOKEN_UNKNOWN                   1
#define LEX_TOKEN_NOTHING                   2
#define LEX_TOKEN_MATHS_OP                  3
#define LEX_TOKEN_LOGICAL_OP                4
#define LEX_TOKEN_KEYWORD                   5
#define LEX_TOKEN_NAME                      6
#define LEX_TOKEN_NUMBER                    7
#define LEX_TOKEN_STRING                    8
#define LEX_TOKEN_OPEN_CURLY                9
#define LEX_TOKEN_CLOSE_CURLY              10

#define LEX_TOKEN_MATHS_OP_PLUS             0
#define LEX_TOKEN_MATHS_OP_MINUS            1
#define LEX_TOKEN_MATHS_OP_TIMES            2
#define LEX_TOKEN_MATHS_OP_DIVIDE           3
#define LEX_TOKEN_MATHS_OP_MOD              4
#define LEX_TOKEN_MATHS_OP_AND              5
#define LEX_TOKEN_MATHS_OP_OR               6
#define LEX_TOKEN_MATHS_OP_XOR              7
#define LEX_TOKEN_MATHS_OP_NOT              8
#define LEX_TOKEN_MATHS_OP_LEFT_SHIFT       9
#define LEX_TOKEN_MATHS_OP_RIGHT_SHIFT     10
#define LEX_TOKEN_MATHS_OP_ASSIGN          11
#define LEX_TOKEN_MATHS_OP_ASSIGN_ADD      12
#define LEX_TOKEN_MATHS_OP_ASSIGN_SUB      13
#define LEX_TOKEN_MATHS_OP_ASSIGN_MUL      14
#define LEX_TOKEN_MATHS_OP_ASSIGN_DIV      15
#define LEX_TOKEN_MATHS_OP_POINTER_ASSIGN  16
#define LEX_TOKEN_MATHS_OP_POINTER_DEREF   17
#define LEX_TOKEN_MATHS_OP_PLUSPLUS        18
#define LEX_TOKEN_MATHS_OP_MINUSMINUS      19

#define LEX_TOKEN_LOGICAL_OP_LT             0
#define LEX_TOKEN_LOGICAL_OP_LE             1
#define LEX_TOKEN_LOGICAL_OP_EQ             2
#define LEX_TOKEN_LOGICAL_OP_NE             3
#define LEX_TOKEN_LOGICAL_OP_GE             4
#define LEX_TOKEN_LOGICAL_OP_GT             5
#define LEX_TOKEN_LOGICAL_OP_AND            6
#define LEX_TOKEN_LOGICAL_OP_OR             7
#define LEX_TOKEN_LOGICAL_OP_NOT            8

#define LEX_TOKEN_KEYWORD_DUP               0
#define LEX_TOKEN_KEYWORD_POP               1
#define LEX_TOKEN_KEYWORD_SWAP              2
#define LEX_TOKEN_KEYWORD_VAR               3
#define LEX_TOKEN_KEYWORD_IF                4
#define LEX_TOKEN_KEYWORD_IFTHEN            5
#define LEX_TOKEN_KEYWORD_WHILE             6
#define LEX_TOKEN_KEYWORD_EXECUTE           7
#define LEX_TOKEN_KEYWORD_PRINT             8
#define LEX_TOKEN_KEYWORD_PRINTLN           9
#define LEX_TOKEN_KEYWORD_PRINTF           10
#define LEX_TOKEN_KEYWORD_SPRINTF          11
#define LEX_TOKEN_KEYWORD_TRUE             12
#define LEX_TOKEN_KEYWORD_FALSE            13
#define LEX_TOKEN_KEYWORD_INT              14
#define LEX_TOKEN_KEYWORD_DOUBLE           15
#define LEX_TOKEN_KEYWORD_BREAK            16
#define LEX_TOKEN_KEYWORD_EXIT             17
#define LEX_TOKEN_KEYWORD_VALUE            18
#define LEX_TOKEN_KEYWORD_DEFINED          19
#define LEX_TOKEN_KEYWORD_INITIALISED      20
#define LEX_TOKEN_KEYWORD_TONAME           21
#define LEX_TOKEN_KEYWORD_NAMESPACE        22
#define LEX_TOKEN_KEYWORD_LIBRARY          23
#define LEX_TOKEN_KEYWORD_FUNCTION         24
#define LEX_TOKEN_KEYWORD_RETURN           25
#define LEX_TOKEN_KEYWORD_DEBUG            26
#define LEX_TOKEN_KEYWORD_STACK            27
#define LEX_TOKEN_KEYWORD_BTREE            28

// How the lexical analyser stores things.

// Replacements.
struct LexReplacement {
  char *keyword;
  char *replacement;
  LexReplacement *next;
};

// Holds a number.
struct LexNumber {
  bool intRepresentation;
  INT valueInt;
  double valueDouble;
};

struct Lex {
  int token;
  char unknownToken;
  int mathsOp;
  int logicalOp;
  int keyword;
  LexNumber number;
  char str[128];
};

struct Keyword {
  const char *string;
  int keyword;
};

// Size of the stack to keep track of nested code segments.
#define OL_STACK_SIZE 256
